utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java

```diff
@@ -0,0 +1,307 @@
+
+              if (!startingLocationFeatures.isEmpty()) {
+                Geometry startingLocationGeometry = startingLocationFeatures.get(0).getGeometry();
+
+                if (startingLocationGeometry instanceof Point){
+                  Point startingLocationGeometryPoint = (Point) startingLocationGeometry;
+
+                // create a graphic for the starting location and add it to the graphics overlay
+                Graphic startingLocationGraphic = new Graphic(startingLocationGeometry, startingPointSymbol);
+                startingLocationGraphicsOverlay.getGraphics().add(startingLocationGraphic);
+
+                // set the map's viewpoint to the starting location
+                mapView.setViewpointAsync(new Viewpoint(startingLocationGeometryPoint, 3000));
+
+                // build the choice list for categories populated with the 'Name' property of each 'UtilityCategory' in the 'UtilityNetworkDefinition'
+                categorySelectionComboBox.getItems().addAll(networkDefinition.getCategories());
+                categorySelectionComboBox.getSelectionModel().select(0);
+                categorySelectionComboBox.setCellFactory(param -> new UtilityCategoryListCell());
+                categorySelectionComboBox.setButtonCell(new UtilityCategoryListCell());
+
+                // enable the UI
+                enableUI(true);
+
+                // update the status text
+                statusLabel.setText("Utility network loaded. Ready to perform trace...");
+                }
+
+              } else {
+                new Alert(Alert.AlertType.ERROR, "Error getting starting location geometry.").show();
+              }
+            } catch (ExecutionException | InterruptedException e) {
+              new Alert(Alert.AlertType.ERROR, "Error getting starting location feature.").show();
+            }
+          });
+
+        } else {
+          new Alert(Alert.AlertType.ERROR, "Error loading Utility Network.").show();
+        }
+      });
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+  }
+
+  /**
+   * Uses the starting location and the selected filter barrier category to perform a valve isolation trace, then
+   * selects all connected elements found in the trace to highlight them.
+   */
+  @FXML
+  private void handleTraceClick() {
+    try {
+      // clear previous selection from the layers
+      mapView.getMap().getOperationalLayers().forEach(layer -> {
+        if (layer instanceof FeatureLayer) {
+          ((FeatureLayer) layer).clearSelection();
+        }
+      });
+
+      // disable the UI, show the progress indicator and update the status text
+      enableUI(false);
+      statusLabel.setText("Running isolation trace...");
+
+      // get the selected utility category
+      if (categorySelectionComboBox.getSelectionModel().getSelectedItem() != null) {
+        UtilityCategory selectedCategory = categorySelectionComboBox.getSelectionModel().getSelectedItem();
```

Comment:
Could move this before the `if` and avoid calling `categorySelectionComboBox.getSelectionModel().getSelectedItem()` twice

Reply:
Thanks, updated!